proposals/0013-expanded-comparison-sampling.md

```diff
@@ -0,0 +1,432 @@
+In addition, this represents a discrepancy with comparable platforms where this
+ deficiency makes porting, emulating, or developing for multiple platforms
+ difficult or impossible.
+
+## Proposed solution
+
+To add full comparison sampling support, HLSL will introduce an optional feature
+ that adds these new comparison sampling builtin methods to all texture objects
+ that currently support the non-comparison methods:
+
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias)`
+  performs a comparison sample using the implicitly-determined LOD with the
+  explicitly-provided _Bias_ value added to that result.
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY)`
+   samples using the LOD calculated using the explicitly-provided
+   _DDX_ and _DDY_ gradients.
+
+Only the simplest overloads of the new and existing builtin methods are
+ described in detail here for the sake of brevity,
+Though not included above, all relevant overloads of `SampleBias` and
+ `SampleGrad` will be reflected in their new `SampleCmpBias` and
+ `SampleCmpGrad` counterparts.
+The parameters not explicitly described here will function exactly as they do
+ when used by the non-comparison builtin methods:
+
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias)`
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias, int<num> Offset)`
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias, int<num> Offset, float Clamp)`
+* `SampleCmpBias(SamplerComparisonState S, float<num> Location, float Bias, int<num> Offset, float Clamp, uint Status)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY, int<num> Offset)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY, int<num> Offset, float Clamp)`
+* `SampleCmpGrad(SamplerState S, float<num> Location, float DDX, float DDY, int<num> Offset, float Clamp, uint Status)`
+
+To add LOD calculation for comparison samplers, HLSL will introduce and require
+ these new comparison sampler method overloads to all texture objects that
+ currently support the non-comparison method overloads:
+
+* `CalculateLevelOfDetailUnclamped(SamplerComparisonState S, float<num> Location)`
+  returns the raw LOD value determined implicitly using gradients representing
+  how much the _Location_ changes across neighboring pixels regardless of
+  whether it falls outside the valid range.
+* `CalculateLevelOfDetail(SamplerComparisonState S, float<num> Location)`
+  returns the implicitly-determined LOD value clamped to the valid range.
+
+## Detailed Design
+
+### HLSL additions
+
+These new HLSL texture methods will be added:
+
+```c++
+<Format> <TexObject>::SampleCmpGrad(
+    SamplerComparisonState S,
+    float<num> Location,
+    float<num> DDX,
+    float<num> DDY,
```

Comment:
`num` for these arguments are not all the same.  For the `Location` argument, it may be one dimension higher for `Texture*Array` types.  Therefore, there needs to be two different `num` variables: one that is the count of texture dimensions without the array dimension, and one with the array dimension.

Dimension with array is used for `Location`, while dimension without array is used for `DDX`, `DDY`, and `Offset`.

Reply:
Right. Careless copy and pasting.

After I wrote this, I ended up editing the table in DXIL.rst because it had some inaccuracies for Cube map texture sample methods: https://github.com/microsoft/DirectXShaderCompiler/pull/6131